README.md
# BDK Manager for iOS / Swift

This package makes it easier to work with [bdk-swift](https://github.com/bitcoindevkit/bdk-swift) on iOS by providing good defaults, simple setup and modern SwiftUI compatible convenience functions and variables.  

It is still a work in progress and not ready for production.

## Installation

Add this github repository https://github.com/bdgwallet/bdkmanager-swift as a dependency in your Xcode project.   
You can then import and use the `BDKManager` library in your Swift code.

```swift
import BDKManager
```

## Setup

To initalise a BDKManager you need to tell it what bitcoin `Network` it should use, what `SyncSource` the wallet is going to connect to for blockchain data, and where the `Database` should store information. The two supported sync source types by BDK on iOS at the moment is Esplora and Electrum API servers. You can specify a custom URL to a private server, or if none is supplied it will default to the public [Blockstream APIs](https://github.com/Blockstream/esplora/blob/master/API.md).

```swift
let network = Network.testnet // .bitcoin, .testnet, .signet or .regtest
let syncSource = SyncSource(type: SyncSourceType.esplora, customUrl: nil) // .esplora or .electrum, optional customUrl
let database = Database(type: DatabaseType.memory, path: nil, treeName: nil) // .memory or .disk, optional path and tree parameters
        
bdkManager = BDKManager.init(network: network, syncSource: syncSource, database: database)   
```

## Load wallet

To create a new mnemonic, descriptor and load the wallet:

```swift
do {    
    let wordCount = WordCount.words12 // .words12, .words24
    let mnemonic = generateMnemonic(wordCount: wordCount)
    let descriptorType = DescriptorType.singleKey_tr86 // .singleKey_tr86, .singleKey_wpkh84
    let descriptor = bdkManager.descriptorFromMnemonic(
        descriptorType: descriptorType,
        mnemonic: mnemonic,
        password: nil) // optional password
    bdkManager.loadWallet(descriptor: descriptor)
} catch let error {
    print(error)
}  
```

To load  wallet from an existing descriptor:

```swift
let descriptor = "wpkh(tprv8ZgxMBicQKsPeSitUfdxhsVaf4BXAASVAbHypn2jnPcjmQZvqZYkeqx7EHQTWvdubTSDa5ben7zHC7sUsx4d8tbTvWdUtHzR8uhHg2CW7MT/*)"
bdkManager.loadWallet(descriptor: descriptor) 
```

## Sync

The wallet can either be synced manually by calling `sync()`, or at regular intervals by using `startSyncRegularly` and `stopSyncRegularly`.
On every sync, the @Published parameters `.balance` and `.transactions` are updated, which means they automatically trigger updates in SwiftUI.

```swift
bdkManager.sync() // Will sync once

bdkManager.startSyncRegularly(interval: 120) // Will sync immediately, then every 120 seconds
bdkManager.stopSyncRegularly() // Will stop the regular sync
```

## Example

A working SwiftUI example app is included in the repo. It has very basic functionality of showing the balance for a descriptor address. In this case the bdkManager is an @ObservedObject, which enables the WalletView to automatically update depending on bdkManager.syncState. The two files required:

**WalletApp.swift**
```swift
import SwiftUI
import BDKManager

@main
struct WalletApp: App {
    @ObservedObject var bdkManager: BDKManager
    
    init() {
        // Define BDKManager init options
        let network = Network.testnet // set bitcoin, testnet, signet or regtest
        let syncSource = SyncSource(type: SyncSourceType.esplora, customUrl: nil) // set esplora or electrum, can take customUrl
        let database = Database(type: DatabaseType.memory, path: nil, treeName: nil) // set memory or disk, optional path and tree parameters
        
        // Initialize a BDKManager instance
        bdkManager = BDKManager.init(network: network, syncSource: syncSource, database: database)
        
        // Load a singlekey wallet from a newly generated private key
        do {
            let wordCount = WordCount.words12 // .words12, .words24
            let mnemonic = generateMnemonic(wordCount: wordCount)
            let descriptorType = DescriptorType.singleKey_tr86 // .singleKey_tr86, .singleKey_wpkh84
            let descriptor = bdkManager.descriptorFromMnemonic(
                descriptorType: descriptorType,
                mnemonic: mnemonic,
                password: nil) // optional password
            bdkManager.loadWallet(descriptor: descriptor)
        } catch let error {
            print(error)
        }
        
        // Or load a wallet from an existing descriptor
        //let descriptor = "wpkh(tprv8ZgxMBicQKsPeSitUfdxhsVaf4BXAASVAbHypn2jnPcjmQZvqZYkeqx7EHQTWvdubTSDa5ben7zHC7sUsx4d8tbTvWdUtHzR8uhHg2CW7MT/*)"
        //bdkManager.loadWallet(descriptor: descriptor)
    }
    
    var body: some Scene {
        WindowGroup {
            WalletView()
                .environmentObject(bdkManager)
        }
    }
}
```

**WalletView.swift**
```swift
import SwiftUI
import BDKManager

struct WalletView: View {
    @EnvironmentObject var bdkManager: BDKManager
    
    var body: some View {
        VStack (spacing: 50){
            Text("Hello, wallet!")
            switch bdkManager.syncState {
            case .synced:
                Text("Balance: \(bdkManager.balance.total.description)")
            case .syncing:
                Text("Balance: Syncing")
            default:
                Text("Balance: Not synced")
            }
            Text(bdkManager.wallet?.getAddress(addressIndex: AddressIndex.new) ?? "-")
        }.onAppear {
            bdkManager.sync() // to sync once
            //bdkManager.startSyncRegularly(interval: 120) // to sync immediately, then every 120 seconds
        }.onDisappear {
            //bdkManager.stopSyncRegularly() // if startSyncRegularly was used
        }
    }
}
```

## Public variables

BDK Manager has the following `@Published` public variables, meaning they can be observed and lead to updates in SwiftUI:
```swift
.wallet: Wallet?
.balance: Balance
.transactions: [TransactionDetails]
.walletState // .empty, .loading, .loaded, .failed
.syncState // .empty, .syncing, .synced, .failed
```

## Public functions

BDK Manager has the following public functions:
```swift
init(network: Network, syncSource: SyncSource, database: Database)
descriptorFromMnemonic(descriptorType: DescriptorType, mnemonic: String, password: String?) -> String?
descriptorFromXprv(descriptorType: DescriptorType, xprv: String) -> String
loadWallet(descriptor: String)
sync()
startSyncRegularly(interval: TimeInterval)
stopSyncRegularly()
sendBitcoin(recipient: String, amount: UInt64, feeRate: Float) -> Bool
```

Since the wallet is a BDK Wallet, the corresponding functions are also available on .wallet:
```swift
getAddress(addressIndex: AddressIndex) -> String
getLastUnusedAddress()  -> String
getBalance() throws -> Balance
sign( psbt: PartiallySignedBitcoinTransaction ) throws
getTransactions() throws -> [Transaction]
getNetwork()  -> Network
broadcast( psbt: PartiallySignedBitcoinTransaction ) throws -> Transaction
```


CHANGELOG.md
# Changelog for BDK Manager for iOS / Swift

An attempt to document breaking or major API changes will be made here.

## 0.9.2

Use SqlLite for disk database instead of deprecated SledDb

## 0.9.1

Fix a bug with how descriptor was derived from mnemonic, changes bdk-ffi 0.9.0 that was missed in previous update
                    
## 0.9.0

Update to [bdk-ffi 0.9.0](https://github.com/bitcoindevkit/bdk-ffi/blob/master/CHANGELOG.md) requiring changes to the way new wallets are generated as well as other API changes.

- BDKManager.balance now returns `Balance` instead of `UInt64`
- BDKManager.transactions now returns `[TransactionDetails]` instead of `[Transaction]`
- New: `descriptorFromMnemonic(descriptorType: DescriptorType, mnemonic: String, password: String?) -> String?`
- Renamed: `createDescriptorFromXprv` to `descriptorFromXprv`
- Removed: `generateExtendedKey(wordCount: WordCount?, password: String?) throws -> ExtendedKeyInfo`
- Removed: `Transaction`
- Removed: `ExtendedKeyInfo`





Package.swift
// swift-tools-version:5.5
// The swift-tools-version declares the minimum version of Swift required to build this package.

import PackageDescription

let package = Package(
    name: "BDKManager",
    platforms: [
           .macOS(.v12),
           .iOS(.v15),
        ],
    products: [
        .library(
            name: "BDKManager",
            targets: ["BDKManager"]),
    ],
    dependencies: [
        .package(url: "https://github.com/bitcoindevkit/bdk-swift", .exactItem("0.5.1")),
    ],
    targets: [
        // Targets are the basic building blocks of a package. A target can define a module or a test suite.
        // Targets can depend on other targets in this package, and on products in packages this package depends on.
        .target(
            name: "BDKManager",
            dependencies: [.product(name: "BitcoinDevKit", package: "bdk-swift")],
            path: "Sources"),
    ]
)


Sources/BDKManager/BDKManager.swift
//  Created by Daniel Nordh on 2/17/22.
//
//  The following repos have been useful in the creation of this code:
//  - BDKSwiftSample by FuturePaul, https://github.com/futurepaul/BdkSwiftSample, which in turn was inspired by
//  - BDK Android Demo Wallet by Thunderbiscuit, https://github.com/thunderbiscuit/bitcoindevkit-android-sample-app

import Foundation
import BitcoinDevKit

public class BDKManager: ObservableObject {
    // Public variables
    @Published public var wallet: Wallet?
    @Published public var balance: Balance = Balance(immature: 0, trustedPending: 0, untrustedPending: 0, confirmed: 0, spendable: 0, total: 0)
    @Published public var transactions: [BitcoinDevKit.TransactionDetails] = []
    @Published public var walletState = WalletState.empty {
        didSet {
            switch walletState {
            case .empty:
                print("Wallet is not initialized")
            case .loading:
                print("Wallet is initializing")
            case .loaded(let wallet):
                print("Wallet is initialized")
                self.wallet = wallet
            case .failed(let error):
                print("Error initializing wallet:" + error.localizedDescription)
            }
        }
    }
    @Published public var syncState = SyncState.empty {
        didSet {
            switch syncState {
            case .empty:
                print("Node is not initialized")
            case .syncing:
                print("Node is syncing")
            case .synced:
                print("Node is synced")
                self.getBalance()
                self.getTransactions()
            case .failed(let error):
                print("Error, node syncing failed: " + error.localizedDescription)
            }
        }
    }
    
    // Private variables
    private var network: Network
    private var syncSource: SyncSource
    private var database: Database
    private let bdkQueue = DispatchQueue (label: "bdkQueue", qos: .userInitiated)
    private var syncTimer: Timer?
    
    // Public functions
    // Create a descriptor from a mnemonic
    public func descriptorFromMnemonic(descriptorType: DescriptorType, mnemonic: String, password: String?) -> String? {
        do {
            let rootKey = try DescriptorSecretKey(network: self.network, mnemonic: mnemonic, password: password ?? nil)
            switch descriptorType {
            case .singleKey_wpkh84:
                let path = try DerivationPath(path: "m/84h/1h/0h/0")
                let descriptorSecretKey = try rootKey.derive(path: path)
                return ("wpkh(" + descriptorSecretKey.asString() + ")")
            case .singleKey_tr86:
                let path = try DerivationPath(path: "m/86'/1'/0'/0")
                let descriptorSecretKey = try rootKey.derive(path: path)
                return ("tr(" + descriptorSecretKey.asString() + ")")
            }
        } catch let error {
            print(error)
        }
        return nil
    }
    
    // Create a descriptor from an extended private key
    public func descriptorFromXprv(descriptorType: DescriptorType, xprv: String) -> String {
        switch descriptorType {
        case .singleKey_wpkh84:
            return ("wpkh(" + xprv + "/84'/1'/0'/0/*)")
        case .singleKey_tr86:
            return ("tr(" + xprv + "/86'/1'/0'/0/*)")
        }
    }
    
    // Initialize a BDKManager instance
    public init(network: Network, syncSource: SyncSource, database: Database) {
        self.network = network
        self.syncSource = syncSource
        self.database = database
    }
    
    // Load a wallet from a descriptor
    public func loadWallet(descriptor: String) {
        self.walletState = WalletState.loading
        let databaseConfig = databaseConfig(database: self.database)
        let blockchainConfig = blockchainConfig(network: self.network, syncSource: self.syncSource)
        initializeWallet(descriptor: descriptor, changeDescriptor: nil, network: self.network, databaseConfig: databaseConfig, blockchainConfig: blockchainConfig)
    }
    
    // Sync the loaded wallet once
    public func sync() {
        switch self.walletState {
        case .loaded(let wallet):
            self.syncState = SyncState.syncing
            bdkQueue.async {
                do {
                    let blockchainConfig = self.blockchainConfig(network: self.network, syncSource: self.syncSource)
                    let blockchain = try Blockchain(config: blockchainConfig)
                    try wallet.sync(blockchain: blockchain, progress: nil)
                    DispatchQueue.main.async {
                        self.syncState = SyncState.synced
                    }
                } catch let error {
                    DispatchQueue.main.async {
                        self.syncState = SyncState.failed(error)
                    }
                }
            }
        default:
            print("Could not sync, wallet not initialized")
        }
    }
    
    // Sync the loaded wallet immediately, and then at the specified regular interval
    public func startSyncRegularly(interval: TimeInterval) {
        self.sync()
        self.syncTimer = Timer.scheduledTimer(withTimeInterval: interval, repeats: true){ tempTimer in
            self.sync()
        }
    }
    
    // Stop regular sync of the loaded wallet
    public func stopSyncRegularly() {
        self.syncTimer?.invalidate()
    }
    
    // Send an amount of bitcoin (in sats) to a recipient, optional feeRate
    public func sendBitcoin(recipient: String, amount: UInt64, feeRate: Float) -> Bool {
        if self.wallet != nil {
            do {
                let psbt = try TxBuilder().addRecipient(address: recipient, amount: amount).feeRate(satPerVbyte: feeRate).finish(wallet: self.wallet!)
                let success = try self.wallet!.sign(psbt: psbt)
                let blockchainConfig = self.blockchainConfig(network: self.network, syncSource: self.syncSource)
                let blockchain = try Blockchain(config: blockchainConfig)
                try blockchain.broadcast(psbt: psbt)
                return success
            } catch let error {
                print(error)
                return false
            }
        } else {
            return false
        }
    }
    
    // Private functions
    // Create a BDK BlockchainConfig based on a SyncSource (.esplora or .electrum)
    private func blockchainConfig(network: Network, syncSource: SyncSource) -> BlockchainConfig {
        var blockchainConfig: BlockchainConfig
        switch syncSource.type {
        case .esplora:
            let defaultUrl = network == Network.bitcoin ? ESPLORA_URL_BITCOIN : ESPLORA_URL_TESTNET
            let url = syncSource.customUrl != nil ? syncSource.customUrl : defaultUrl
            let esploraConfig = EsploraConfig.init(baseUrl: url!, proxy: nil, concurrency: nil, stopGap: ESPLORA_STOPGAP, timeout: ESPLORA_TIMEOUT)
            blockchainConfig = BlockchainConfig.esplora(config: esploraConfig)
        case .electrum:
            let defaultUrl = network == Network.bitcoin ? ELECTRUM_URL_BITCOIN : ELECTRUM_URL_TESTNET
            let url = syncSource.customUrl != nil ? syncSource.customUrl : defaultUrl
            let electrumConfig = ElectrumConfig(url: url!, socks5: nil, retry: ELECTRUM_RETRY, timeout: nil, stopGap: ELECTRUM_STOPGAP)
            blockchainConfig = BlockchainConfig.electrum(config: electrumConfig)
        }
        return blockchainConfig
    }
    
    // Create a BDK DatabaseConfig based on a Database (.memory or .disk)
    private func databaseConfig(database: Database) -> DatabaseConfig {
        var databaseConfig: DatabaseConfig
        switch database.type {
        case .memory:
            databaseConfig = DatabaseConfig.memory
        case .disk:
            let path = database.path != nil ? database.path : ""
            let sqlLiteDbConfig = SqliteDbConfiguration(path: path!)
            databaseConfig = DatabaseConfig.sqlite(config: sqlLiteDbConfig)
        }
        return databaseConfig
    }
    
    // Initialize a BDK Wallet based on Descriptor, Network, DatabaseConfig and BlockchainConfig
    private func initializeWallet(descriptor: String, changeDescriptor: String?, network: Network, databaseConfig: DatabaseConfig, blockchainConfig: BlockchainConfig) {
        do {
            let wallet = try Wallet.init(descriptor: descriptor, changeDescriptor: changeDescriptor, network: network, databaseConfig: databaseConfig)
            self.walletState = WalletState.loaded(wallet)
        } catch let error {
            self.walletState = WalletState.failed(error)
        }
    }
    
    // Update .balance
    private func getBalance() {
        do {
            self.balance = try self.wallet!.getBalance()
            print("Balance is: " + self.balance.total.description)
        } catch let error {
            print("Error getting wallet balance: " + error.localizedDescription)
        }
    }
    
    // Update .transactions
    private func getTransactions() {
        do {
            let transactions = try self.wallet!.listTransactions()
            self.transactions = transactions
            print("Transaction count: " + self.transactions.count.description)
        } catch let error {
            print("Error getting transactions: " + error.localizedDescription)
        }
    }
}

// Helpers
public typealias Network = BitcoinDevKit.Network
public typealias PartiallySignedBitcoinTransaction = BitcoinDevKit.PartiallySignedBitcoinTransaction
public typealias Progress = BitcoinDevKit.Progress
public typealias TransactionDetails = BitcoinDevKit.TransactionDetails
public typealias WordCount = BitcoinDevKit.WordCount
public typealias AddressIndex = BitcoinDevKit.AddressIndex
public typealias AddressInfo = BitcoinDevKit.AddressInfo

public enum DescriptorType {
    case singleKey_wpkh84 // Native Segwit
    case singleKey_tr86 // Taproot
}

public struct SyncSource {
    public let type: SyncSourceType
    public let customUrl: String?
    
    public init(type: SyncSourceType, customUrl: String?) {
        self.type = type
        self.customUrl = customUrl
    }
}

public enum SyncSourceType {
    case esplora
    case electrum
}

public struct Database {
    public let type: DatabaseType
    public let path: String?
    public let treeName: String?
    
    public init(type: DatabaseType, path: String?, treeName: String?) {
        self.type = type
        self.path = path
        self.treeName = treeName
    }
}

public enum DatabaseType {
    case memory
    case disk
}

public enum SyncState {
    case empty
    case syncing
    case synced
    case failed(Error)
}

public enum WalletState {
    case empty
    case loading
    case loaded(Wallet)
    case failed(Error)
}

// Public API URLs
let ESPLORA_URL_BITCOIN = "https://blockstream.info/api/"
let ESPLORA_URL_TESTNET = "https://blockstream.info/testnet/api"

let ELECTRUM_URL_BITCOIN = "ssl://electrum.blockstream.info:60001"
let ELECTRUM_URL_TESTNET = "ssl://electrum.blockstream.info:60002"

// Defaults
let ESPLORA_TIMEOUT = UInt64(1000)
let ESPLORA_STOPGAP = UInt64(20)

let ELECTRUM_RETRY = UInt8(5)
let ELECTRUM_STOPGAP = UInt64(10)


Sources/BDGWalletBDK/BDGWalletBDK.swift
//  Created by Daniel Nordh on 2/17/22.
//
//  The following repos have been useful in the creation of this code:
//  - BDKSwiftSample by FuturePaul, https://github.com/futurepaul/BdkSwiftSample, which in turn was inspired by
//  - BDK Android Demo Wallet by Thunderbiscuit, https://github.com/thunderbiscuit/bitcoindevkit-android-sample-app

import Foundation
import BitcoinDevKit

public class BDGWalletBDK: ObservableObject {
    @Published public var wallet: Wallet?
    @Published public var balance: UInt64 = 0
    @Published public var transactions: [BitcoinDevKit.Transaction] = []
    @Published public var walletState = WalletState.empty {
        didSet {
            switch walletState {
            case .empty:
                print("Wallet is not initialized")
            case .initializing:
                print("Wallet is initializing")
            case .initialized(let wallet):
                print("Wallet is initialized")
                self.wallet = wallet
                self.sync()
            case .failed(let error):
                print("Error initializing wallet:" + error.localizedDescription)
            }
        }
    }
    @Published public var syncState = SyncState.empty {
        didSet {
            switch syncState {
            case .empty:
                print("Node is not initialized")
            case .syncing:
                print("Node is syncing")
            case .synced:
                print("Node is synced")
                self.getBalance()
                self.getTransactions()
            case .failed(let error):
                print("Error, node syncing failed: " + error.localizedDescription)
            }
        }
    }
    
    private let bdkQueue = DispatchQueue (label: "bdkQueue", qos: .userInitiated)
    
    public init(descriptor: String, network: Network, syncSource: SyncSource, database: Database) {
        self.walletState = WalletState.initializing
        let databaseConfig = databaseConfig(database: database)
        let blockchainConfig = blockchainConfig(network: network, syncSource: syncSource)
        initializeWallet(descriptor: descriptor, changeDescriptor: nil, network: network, databaseConfig: databaseConfig, blockchainConfig: blockchainConfig)
    }
    
    private func blockchainConfig(network: Network, syncSource: SyncSource) -> BlockchainConfig {
        var blockchainConfig: BlockchainConfig
        switch syncSource.type {
        case .esplora:
            let defaultUrl = network == Network.bitcoin ? ESPLORA_URL_BITCOIN : ESPLORA_URL_TESTNET
            let url = syncSource.customUrl != nil ? syncSource.customUrl : defaultUrl
            let esploraConfig = EsploraConfig.init(baseUrl: url!, proxy: nil, timeoutRead: 1000, timeoutWrite: 1000, stopGap: 20)
            blockchainConfig = BlockchainConfig.esplora(config: esploraConfig)
        case .electrum:
            let defaultUrl = network == Network.bitcoin ? ELECTRUM_URL_BITCOIN : ELECTRUM_URL_TESTNET
            let url = syncSource.customUrl != nil ? syncSource.customUrl : defaultUrl
            let electrumConfig = ElectrumConfig(url: url!, socks5: nil, retry: 5, timeout: nil, stopGap: 10)
            blockchainConfig = BlockchainConfig.electrum(config: electrumConfig)
        }
        return blockchainConfig
    }
    
    private func databaseConfig(database: Database) -> DatabaseConfig {
        var databaseConfig: DatabaseConfig
        switch database.type {
        case .memory:
            databaseConfig = DatabaseConfig.memory(junk: "")
        case .disk:
            let path = database.path != nil ? database.path : ""
            let treeName = database.treeName != nil ? database.treeName : ""
            let sledDbConfig = SledDbConfiguration(path: path!, treeName: treeName!)
            databaseConfig = DatabaseConfig.sled(config: sledDbConfig)
        }
        return databaseConfig
    }
    
    private func initializeWallet(descriptor: String, changeDescriptor: String?, network: Network, databaseConfig: DatabaseConfig, blockchainConfig: BlockchainConfig) {
        do {
            let wallet = try Wallet.init(descriptor: descriptor, changeDescriptor: nil, network: network, databaseConfig: databaseConfig, blockchainConfig: blockchainConfig)
            self.walletState = WalletState.initialized(wallet)
        } catch let error {
            self.walletState = WalletState.failed(error)
        }
    }
    
    public func sync() {
        switch self.walletState {
        case .initialized(let wallet):
            self.syncState = SyncState.syncing
            bdkQueue.async {
                do {
                    let progress = Progress() // Progress for Esplora/Electrum is not available, but object is required by the wallet.sync function.
                    try wallet.sync(progressUpdate: progress, maxAddressParam: nil)
                    DispatchQueue.main.async {
                        self.syncState = SyncState.synced
                    }
                } catch let error {
                    DispatchQueue.main.async {
                        self.syncState = SyncState.failed(error)
                    }
                }
            }
        default:
            print("Could not sync, wallet not initialized")
        }
    }
    
    private func getBalance() {
        do {
            self.balance = try self.wallet!.getBalance()
            print("Balance is: " + self.balance.description)
        } catch let error {
            print("Error getting wallet balance: " + error.localizedDescription)
        }
    }
    
    private func getTransactions() {
        do {
            let transactions = try self.wallet!.getTransactions()
            self.transactions = transactions.sorted(by: {
                switch $0 {
                case .confirmed(_, let confirmation_a):
                    switch $1 {
                    case .confirmed(_, let confirmation_b): return confirmation_a.timestamp > confirmation_b.timestamp
                    default: return false
                    }
                default:
                    switch $1 {
                    case .unconfirmed(_):
                        return true
                    default: return false
                    }
                }
            })
            print("Transaction count: " + self.transactions.count.description)
        } catch let error {
            print("Error getting transactions: " + error.localizedDescription)
        }
    }
}

// Structs, Classes and enums

public typealias Network = BitcoinDevKit.Network

public struct SyncSource {
    public let type: SyncSourceType
    public let customUrl: String?
    
    public init(type: SyncSourceType, customUrl: String?) {
        self.type = type
        self.customUrl = customUrl
    }
}

public enum SyncSourceType {
    case esplora
    case electrum
}

public struct Database {
    public let type: DatabaseType
    public let path: String?
    public let treeName: String?
    
    public init(type: DatabaseType, path: String?, treeName: String?) {
        self.type = type
        self.path = path
        self.treeName = treeName
    }
}

public enum DatabaseType {
    case memory
    case disk
}

public enum SyncState {
    case empty
    case syncing
    case synced
    case failed(Error)
}

public enum WalletState {
    case empty
    case initializing
    case initialized(Wallet)
    case failed(Error)
}

class Progress : BdkProgress {
    var value = Float()
    func update(progress: Float, message: String?) {
        value = progress
        print("progress", progress, message as Any)
    }
}

// Defeault public api URLs

let ESPLORA_URL_BITCOIN = "https://blockstream.info/api/"
let ESPLORA_URL_TESTNET = "https://blockstream.info/testnet/api"

let ELECTRUM_URL_BITCOIN = "ssl://electrum.blockstream.info:60001"
let ELECTRUM_URL_TESTNET = "ssl://electrum.blockstream.info:60002"



Example/BDKManagerExampleApp/BDKManagerExampleApp.swift
//
//  BDKManagerExampleApp.swift
//  BDKManagerExampleApp
//
//  Created by Daniel Nordh on 3/18/22.
//

import SwiftUI
import BDKManager

@main
struct BDKManagerExampleApp: App {
    @ObservedObject var bdkManager: BDKManager
    
    init() {
        // Define BDKManager init options
        let network = Network.testnet // .bitcoin, .testnet, .signet or .regtest
        let syncSource = SyncSource(type: SyncSourceType.esplora, customUrl: nil) // .esplora or .electrum, can take customUrl
        let database = Database(type: DatabaseType.memory, path: nil, treeName: nil) // .memory or .disk, optional path and tree parameters
        
        // Initialize a BDKManager instance
        bdkManager = BDKManager.init(network: network, syncSource: syncSource, database: database)
        
        // Load a singlekey wallet from a newly generated mnemonic
        do {
            let wordCount = WordCount.words12 // .words12, .words24
            let mnemonic = generateMnemonic(wordCount: wordCount)
            let descriptorType = DescriptorType.singleKey_tr86 // .singleKey_tr86, .singleKey_wpkh84
            let descriptor = bdkManager.descriptorFromMnemonic(
                descriptorType: descriptorType,
                mnemonic: mnemonic,
                password: nil)
            bdkManager.loadWallet(descriptor: descriptor)
        } catch let error {
            print(error)
        }
        
        // Or load a wallet from an existing descriptor
        //let descriptor = "wpkh(tprv8ZgxMBicQKsPeSitUfdxhsVaf4BXAASVAbHypn2jnPcjmQZvqZYkeqx7EHQTWvdubTSDa5ben7zHC7sUsx4d8tbTvWdUtHzR8uhHg2CW7MT/*)"
        //bdkManager.loadWallet(descriptor: descriptor)
    }
    
    var body: some Scene {
        WindowGroup {
            WalletView()
                .environmentObject(bdkManager)
        }
    }
}


Example/BDKManagerExampleApp/WalletView.swift
//
//  WalletView.swift
//  BDKManagerExampleApp
//
//  Created by Daniel Nordh on 3/18/22.
//

import SwiftUI
import BDKManager

struct WalletView: View {
    @EnvironmentObject var bdkManager: BDKManager
    
    var body: some View {
        VStack (spacing: 50){
            Text("Hello, wallet!")
            switch bdkManager.syncState {
            case .synced:
                Text("Balance: \(bdkManager.balance.total.description)")
            case .syncing:
                Text("Balance: Syncing")
            default:
                Text("Balance: Not synced")
            }
            Text(bdkManager.wallet?.getAddress(addressIndex: AddressIndex.new) ?? "-")
        }.onAppear {
            bdkManager.sync() // to sync once
            //bdkManager.startSyncRegularly(interval: 120) // to sync every 120 seconds
        }.onDisappear {
            //bdkManager.stopSyncRegularly() // if startSyncRegularly was used
        }
    }
}
